packages/contracts/contracts/FlippenZap.sol

```diff
@@ -0,0 +1,248 @@
+        } else {
+            _expectedDebt = _calculateCdpDebtForZap(_expectedColl, _leverage);
+            require(
+                ebtcCollToken.balanceOf(address(this)) >= _expectedColl,
+                "FlippenZap: not enough collateral to openCdpFor()"
+            );
+            _newCdp = ebtcBorrowerOperations.openCdpFor(
+                _expectedDebt,
+                bytes32(0),
+                bytes32(0),
+                _expectedColl,
+                msg.sender
+            );
+
+            // transfer minted eBTC
+            // TODO do we really need to swap minted eBTC here on caller's behalf?
+            require(
+                ebtcToken.balanceOf(address(this)) >= _expectedDebt,
+                "FlippenZap: no expected debt after openCdpFor()"
+            );
+            ebtcToken.transfer(msg.sender, _expectedDebt);
+        }
+        return _newCdp;
+    }
+
+    function _calculateCdpDebtForZap(uint256 _coll, uint256 _leverage) internal returns (uint256) {
+        uint256 denominator = _leverage > 1 ? 1e18 : (CCR + ICR_SMALL_BUFFER);
+        uint256 _price = ebtcPriceFeed.fetchPrice();
+        return (_coll * _price) / denominator;
+    }
+
+    function _singleSwapViaBalancer(
+        address tokenIn,
+        address tokenOut,
+        uint256 amountIn,
+        uint256 expectedOut
+    ) internal returns (uint256) {
+        SingleSwap memory singleSwap = SingleSwap(
+            balancerPoolId,
+            SwapKind.GIVEN_IN,
+            tokenIn,
+            tokenOut,
+            amountIn,
+            ""
+        );
+        FundManagement memory funds = FundManagement(address(this), false, address(this), false);
+        return balancerVault.swap(singleSwap, funds, expectedOut, block.timestamp);
+    }
+
+    /// @inheritdoc IERC3156FlashBorrower
+    function onFlashLoan(
+        address initiator,
+        address token,
+        uint256 amount,
+        uint256 fee,
+        bytes calldata data
+    ) external override returns (bytes32) {
+        require(initiator == address(this), "FlippenZap: flashloan has to be initiated from this");
+        require(
+            msg.sender == address(ebtcBorrowerOperations),
+            "FlippenZap: flashloan call has to be from BorrowerOperations"
+        );
+        require(token == address(ebtcToken), "FlippenZap: flashloan token has to be eBTC");
+        require(amount > 0, "FlippenZap: flashloan amount has to be above zero");
+
+        // decode from flashloan parameter for new leveraged CDP
+        (address _borrower, uint256 _initialColl, uint256 _collToLoan) = abi.decode(
+            data,
+            (address, uint256, uint256)
+        );
+        uint256 _expectedDebt = amount + fee;
+
+        // swap eBTC to collateral
+        uint256 _outAmount = _singleSwapViaBalancer(
+            address(ebtcToken),
+            address(ebtcCollToken),
+            amount,
+            ((_collToLoan * (MAX_SLIPPAGE - slippage)) / MAX_SLIPPAGE)
```

Comment:
I think slippage should be a parameter of `enterLongEth`

Reply:
We may want to hide the implementation details (like the flashloan involved) from the end user so UI could focus clearly on convenient one-click Zap-in for `enterLongEth`

@dapp-whisperer please share your insight on this.